### ISSUE TITLE
elidable_lifetime_names doesn't understand conditional compilation

### ISSUE THREAD

**Issue by nazar-pc** [rust-lang/rust-clippy#17638]
### Summary

Example in a real project: https://github.com/nazar-pc/abundance/blob/0ed3edab729a715abfb31a529bda0ef154817202/crates/contracts/core/ab-contracts-common/src/env.rs#L116-L129

### Lint Name

elidable_lifetime_names

### Reproducer

I tried this code:

```rust
#[repr(C)]
pub struct Env<'a> {
    state: EnvState,
    #[cfg(feature = "executor")]
    executor_context: &'a mut dyn ExecutorContext,
    #[cfg(not(feature = "executor"))]
    phantom_data: PhantomData<&'a ()>,
}

impl<'a> Env<'a> {
    #[cfg(feature = "executor")]
    pub fn with_executor_context(
        state: EnvState,
        executor_context: &'a mut dyn ExecutorContext,
    ) -> Self {
        Self {
            state,
            executor_context,
        }
    }
}
```

I saw this happen:

```
warning: the following explicit lifetimes could be elided: 'a
   --> crates/contracts/core/ab-contracts-common/src/env.rs:116:6
    |
116 | impl<'a> Env<'a> {
    |      ^^      ^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/main/index.html#elidable_lifetime_names
    = note: `-W clippy::elidable-lifetime-names` implied by `-W clippy::pedantic`
    = help: to override `-W clippy::pedantic` add `#[allow(clippy::elidable_lifetime_names)]`
help: elide the lifetimes
    |
116 - impl<'a> Env<'a> {
116 + impl Env<'_> {
    |
```

I expected to see this happen: No lint since without that lifetime the code doesn't compile under a different set of features:
```
error: lifetime may not live long enough
   --> crates/contracts/core/ab-contracts-common/src/env.rs:125:9
    |
123 |           executor_context: &'_ mut dyn ExecutorContext,
    |                             - let's call the lifetime of this reference `'1`
124 |       ) -> Self {
    |            ---- return type is env::Env<'2>
125 | /         Self {
126 | |             state,
127 | |             executor_context,
128 | |         }
    | |_________^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`
    |
    = note: requirement occurs because of the type `env::Env<'_>`, which makes the generic argument `'_` invariant
    = note: the struct `env::Env<'a>` is invariant over the parameter `'a`
    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
```

### Version

```text
rustc 1.100.0-nightly (e7769602a 2026-08-24)
binary: rustc
commit-hash: e7769602aca3770e8d8ea55716becb22e839a579
commit-date: 2026-08-24
host: x86_64-unknown-linux-gnu
release: 1.100.0-nightly
LLVM version: 23.1.0
```

**Comment by CommanderStorm** [comment]
I don't think this is a false positive, is it?
If you cfg to !executor, the lifetime name CAN be elided...

**Comment by nazar-pc** [comment]
I guess it could be argued both ways. On one hand lifetimes can't be conditionally compiled like methods, so you can't really make it lint with and without that feature successfully at the same time. On the other hand, in this specific case, it is possible to split `impl Env` into two: one with explicit lifetime and one without (real code has more methods, I only included one in the example for brevity).

**Comment by Jarcho** [comment]
This is not going to be fixed. We can't see any code removed via `cfg`, nor can we reasonable detect it in general (it could be anywhere in the impl block, not just at the item level). Even if we did detect it we still couldn't check whether it's used in those blocks since that would require re-parsing and running name resolution. This also behaves the same way all of rustc's unused lints work were they consider only the current compilation after disabled code is removed.

The easiest workaround here would be to move all the `cfg(...)` items to their own impl block and lift the `cfg` up to that level. You could also add `#[cfg_attr(not(...), expect(lint))]`

**Comment by Jarcho** [comment]
If the compiler allowed `cfg` on generic arguments you could also use go that route. Generic declarations are allowed to have them (`impl<#[cfg(..) 'a> Foo {}` compiles fine) and function arguments also work so I don't see any reason the compiler should reject it on generic arguments, it just currently does.

**Comment by CommanderStorm** [comment]
> I don't see any reason the compiler should reject it on generic arguments, it just currently does

readability? 👀 Seems like something fairly confusing to read...
